src/Microsoft.AspNetCore.Mvc.Core/ApplicationParts/RelatedAssemblyAttribute.cs

```diff
@@ -2,23 +2,110 @@
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
 using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Reflection;
 using Microsoft.AspNetCore.Mvc.Core;
 
 namespace Microsoft.AspNetCore.Mvc.ApplicationParts
 {
+    /// <summary>
+    /// Specifies a assembly to load as part of MVC's assembly discovery mechanism.
+    /// </summary>
     [AttributeUsage(AttributeTargets.Assembly, AllowMultiple = true)]
     public sealed class RelatedAssemblyAttribute : Attribute
     {
-        public RelatedAssemblyAttribute(string name)
+        private static readonly Func<string, Assembly> AssemblyLoadFileDelegate = Assembly.LoadFile;
+
+        /// <summary>
+        /// Initializes a new instance of <see cref="RelatedAssemblyAttribute"/>.
+        /// </summary>
+        /// <param name="assemblyFileName">The file name, without extension, of the related assembly.</param>
+        public RelatedAssemblyAttribute(string assemblyFileName)
         {
-            if (string.IsNullOrEmpty(name))
+            if (string.IsNullOrEmpty(assemblyFileName))
             {
-                // Temporary workaround until we have a new build of RazorSdk.
-                // TODO: Uncomment the below line.
-                // throw new ArgumentException(Resources.ArgumentCannotBeNullOrEmpty, nameof(name));
+                throw new ArgumentException(Resources.ArgumentCannotBeNullOrEmpty, nameof(assemblyFileName));
             }
+
+            AssemblyFileName = assemblyFileName;
         }
 
-        public string Name { get; }
+        /// <summary>
+        /// Gets the assembly file name without extension.
+        /// </summary>
+        public string AssemblyFileName { get; }
+
+        /// <summary>
+        /// Gets <see cref="Assembly"/> instances specified by <see cref="RelatedAssemblyAttribute"/>.
+        /// </summary>
+        /// <param name="assembly">The assembly containing <see cref="RelatedAssemblyAttribute"/> instances.</param>
+        /// <param name="throwOnError">Determines if the method throws if a related assembly could not be located.</param>
+        /// <returns>Related <see cref="Assembly"/> instances.</returns>
+        public static IReadOnlyList<Assembly> GetRelatedAssemblies(Assembly assembly, bool throwOnError)
```

Comment:
When would you not want to throw?

Reply:
It's detailed in the other block. You may have `RelatedAssemblyPartAttribute` without a corresponding part.

Reply:
> It's detailed in the other block. You may have RelatedAssemblyPartAttribute without a corresponding part.

I still don't understand what you mean

Reply:
Sorry, I thought I had responded to this in a different place. In the event of Razor build, you can set `RazorCompileOnBuild=false`, but `RazorCompileOnPublish=true`. We tell the Sdk to generate the `RelatedAssemblyAttribute` as part of `Build`. However, it's non-trivial in msbuild to determine what target will eventually get run. Consequently you can stamp the assembly with the attribute without producing a related part.